pydantic/main.py

```diff
@@ -57,6 +64,97 @@
+
+    # If value is already a BaseModel, just return that
+    # NOTE: if we see this, we assume that all of the children of the supplied BaseModel have been recursed through and
+    # converted to BaseModels, OR that the user manually provided this value and would like us to *not* recurse through
+    # its children and instead leave it "as-is". Depending on the circumstance, the user might want either behavior, so
+    # perhaps this could be toggled with another parameter/setting.
+    if isinstance(value, BaseModel):
+        return value
+
+    # Try to get a "parent" class if we're dealing with a structural type like list, dict, sequence, etc.
+    # Otherwise just treat the annotation itself as the parent
+    origin = get_origin(annotation)
+    origin = annotation if origin is None else origin
+
+    # Get all of the classes in the brackets, if there are any
+    args = get_args(annotation)
+
+    # Union and Literal don't play well with isinstance/issubclass, so we check them early
+    if origin is Literal:
+        return value
+    elif _is_union(origin):
+        return _handle_union(args, value)
+    # Then we check to see if the annotation describes it as a Base/RootModel
+    elif issubclass(origin, BaseModel):
+        if origin.__pydantic_root_model__:  # my saviour
+            return origin.model_construct(value, _recursive=True)  # treat like RootModel
+        else:
+            try:
+                return origin.model_construct(**value, _recursive=True)  # treat like BaseModel
+            except TypeError:
+                return value
+    # Then we check for subclasses of mappings/sequences/sets
+    # But if there are no arguments to the annotation, we have to treat the value generically anyway
+    elif len(args) == 0:
+        return value
+    elif issubclass(origin, typing.Tuple):
+        if isinstance(args[-1], type(Ellipsis)):  # tuple[T, ...]
+            member_type = args[0]
+            return type(value)(_recursive_model_construct(member_type, v) for v in value)
+        else:  # tuple[A, B, C]
+            return type(value)(_recursive_model_construct(member_type, value[i]) for i, member_type in enumerate(args))
+    elif issubclass(origin, typing.Sequence) and not issubclass(origin, (str, bytes)):
+        member_type = args[0]
+        return [_recursive_model_construct(member_type, v) for v in value]
+    elif issubclass(origin, typing.Mapping):
+        # NOTE: do we want to recurse key? Key must be hashable, and by default BaseModels are not; furthermore, dumping
+        # a model to a dict makes it unhashable even if you implement __hash__ on the model itself. However, there might
+        # be a few crazy cases where people both override __hash__ and  overwrite pydantic's serialization so that it
+        # outputs to a 'HashableDict' or similar, in which case this might be desired... but we could also just wait
+        # until such a feature is required before implementing it.
```

Comment:
I suppose the real question is just "is making a `BaseModel` hashable condoned by Pydantic?" 

If an instance of `BaseModel` isn't hashable, then attempting to `model_construct()` it into a set or a dict key will fail, clearly and predictably. But if it does implement `__hash__`, then currently it will work fine (you wont be able to dump unless you override what it serializes to, but I'm not familiar enough with the internals to know exactly how to do that. You can do that, right?) Point being, as far as I can tell there's nothing stopping use from making it work, so the question becomes "should we bother to make it work?" It would make sense if a model with a field `set[SomeModel]` would get properly constructed into a `set` of `SomeModel`s.

If making a BaseModel hashable is something condoned (and better still if frequently done), then we should probably include the functionality. If not, I don't feel bad ignoring sets and exploding dict keys under the pretense that BaseModels are likely not going to be hashable objects.